ARTEMIS-4532 MQTT-to-core wildcard conversion is broken

jbertram opened the pull request:
Currently when an MQTT topic filter contains characters from the configured wildcard syntax the conversion to/from this syntax breaks.

For example, when using the default wildcard syntax if an MQTT topic filter contains a . the conversion from the MQTT wildcard syntax to the core wildcard syntax and back will result in the `.` being replaced with a `/.`.

This commit fixes that plus a few other things...

 - Implements proper conversions to/from one WildcardConfiguration to another.
 - Refactors the MQTT code which invokes these conversion methods. This includes simplifying a lot of test code.
 - Adds lots of tests for everything.
 - Clarifies some variable naming to better distinguish between core and MQTT.

gemmellr commented on docs/user-manual/versions.adoc:
```diff
@@ -31,6 +31,13 @@ These settings are now no longer required.
 
+* Due to https://issues.apache.org/jira/browse/ARTEMIS-4532[ARTEMIS-4532] the names of addresses and queues related to MQTT topics and subscriptions respectively may change. This will only impact you:
++
+  . If the broker is configured to use a xref:wildcard-syntax.adoc[wildcard syntax] which doesn't match the xref:mqtt.adoc#wildcard-syntax[MQTT wildcard syntax].
+  . If you are using characters from the broker's wildcard syntax in your MQTT topic names or topic filters.
```
It could be clearer from the text both items in the list need apply, just to drive it home

jbertram replied:
Updated.

gemmellr commented on docs/user-manual/versions.adoc:
```diff
@@ -31,6 +31,13 @@ These settings are now no longer required.
+  . If the broker is configured to use a xref:wildcard-syntax.adoc[wildcard syntax] which doesn't match the xref:mqtt.adoc#wildcard-syntax[MQTT wildcard syntax].
+  . If you are using characters from the broker's wildcard syntax in your MQTT topic names or topic filters.
++
+In this case the characters from the broker's wildcard syntax that do not match the characters in the MQTT wildcard syntax will be escaped with a backslash (i.e. `\`).
```
Suggestion for what a user might need to do to adapt? E.g rename an existing queues to what they 'should be' now?

jbertram replied:
Updated.

gemmellr commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/mqtt5/spec/controlpackets/PublishTests.java:
```diff
@@ -311,15 +312,15 @@ public void testRetainFlagFalse() throws Exception {
       producer.close();
 
-      Wait.assertFalse(() -> server.locateQueue(MQTTUtil.convertMqttTopicFilterToCore(MQTTUtil.MQTT_RETAIN_ADDRESS_PREFIX, TOPIC, MQTTUtil.MQTT_WILDCARD)).getMessageCount() > 1, 1000, 100);
+      Wait.assertFalse(() -> getRetainedMessageQueue(TOPIC).getMessageCount() > 1, 1000, 100);
```
Is there a reason why the assert should not just be checking for a count of 2 ?

jbertram replied:
If we just check for 2 and the count is somehow 3 or more then the check will pass even though the behavior is wrong.

gemmellr replied:
Makes sense. I think maybe I misread that when I commented, not sure I saw/considered the wait. I'd generally try to avoid burning 1 second like that by doing something else instead thats likely to be more timely but gives the same 'processing should have occurred' result , e.g publishing another message somewhere else after the one currently being sent here, and waiting for it to arrive (after processing of the earlier message completes) before then checking on the retained message queue. Its only a second, yes, but it quickly adds up...there are ~20,000 tests after all.

gemmellr commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/mqtt5/spec/controlpackets/PublishTests.java:
```diff
@@ -269,15 +270,15 @@ public void testRetainFlagWithEmptyMessage() throws Exception {
       producer.publish(TOPIC, "retain1".getBytes(), 2, true);
 
-      Wait.assertTrue(() -> server.locateQueue(MQTTUtil.convertMqttTopicFilterToCore(MQTTUtil.MQTT_RETAIN_ADDRESS_PREFIX, TOPIC, MQTTUtil.MQTT_WILDCARD)).getMessageCount() == 1, 2000, 100);
+      Wait.assertTrue(() -> getRetainedMessageQueue(TOPIC).getMessageCount() == 1, 2000, 100);
```
Lots of tests asserting the queue count is 1 now, but would be better if they also verified it was 0 to begin with.

jbertram replied:
Fixed.

gemmellr reviewed (commented):
Also needs a rebase + retest now.

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/config/WildcardConfiguration.java:
```diff
@@ -119,19 +138,94 @@ public String getSingleWordString() {
+   }
+
+   private boolean isEscaped(final String input) {
+      for (int i = 0; i < input.length() - 1; i++) {
+         if (input.charAt(i) == ESCAPE && (input.charAt(i + 1) == getDelimiter() || input.charAt(i + 1) == getSingleWord() || input.charAt(i + 1) == getAnyWords())) {
+            return true;
+         }
+      }
```
Shouldn't the `input.length() - 1` mitigate this?

gemmellr replied:
Doesn't seem like it. That governs that the variable i cant be beyond the 0-based string index used for _charAt(i)_, but the loop then also uses _charAt(i+1)_, so for the last character that would become input.length() and be beyond the limit?

jbertram replied:
I think the `i < input.length() - 1` is preventing any issue here. If it was using `<=` or just `input.length()` there would be a problem, but as it stands now you can pass a `String` where the last character is the escape character (e.g. `foo\\`) and there's no IOOBE.

gemmellr replied:
Yes I think you are right, think I misread that too and believed the loop itself was looking at the last char...so it wont throw because the loop never actually looks at the last character, just the if through the charAt(i+1).

jbertram commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/mqtt5/spec/controlpackets/PublishTests.java:
```diff
@@ -653,6 +654,10 @@ public void messageArrived(String topic, MqttMessage message) throws Exception {
       consumer.close();
    }
 
+   private Queue getRetainedMessageQueue(String TOPIC) {
+      return server.locateQueue(MQTTUtil.convertMqttTopicFilterToCore(MQTTUtil.MQTT_RETAIN_ADDRESS_PREFIX, TOPIC, server.getConfiguration().getWildcardConfiguration()));
+   }
```
Can you give me an example of what you mean?

jbertram replied:
BTW, I fixed the parameter name.

gemmellr replied:
This method seems to have been the source of multiple breaks, or changes in behaviour, over the last several releases. Obviously bugs happen, but it feels like some of these would have been caught by comprehensive unit testing of the method. The broker calls the method to name its queues, but the system tests also call the method to get the queue name the broker used. If for example the method returns "wrong-value" as it effectively has been doing, then broker names things in that way, and the system tests also just use that as the expected value, and so in those cases theyll always match and wont notice if anything is wrong, e.g I go in there later and mess something up with the naming.

If on the other hand there is a direct unit testing of the method comprehensively checking the output, ensuring that the exact value expected is returned, then the behaviour of that method is then more directly tested (and a few ms should tell if it is later broken...vs running minutes of system tests that mostly wont notice because they use the same method to establish what is expected), and likely more understandable because there are now specific tests for it showing what its expected to do, and quickly pointing out if changed inappropriately.

jbertram commented on artemis-server/src/test/java/org/apache/activemq/artemis/core/config/WildcardConfigurationTest.java:
```diff
@@ -0,0 +1,40 @@
+import org.junit.Assert;
+import org.junit.Test;
+
+public class WildcardConfigurationTest extends Assert {
```
I added a new test.

gemmellr replied:
The equals method itself should be far better tested in addition to the handling in convert(). Negative tests, tests of varying each applicable component field, checking that the equals and hashCode mutual contract is respected (I caught one in a PR just last week where that was broken).

jbertram replied:
New tests added.

jbertram commented on artemis-server/src/test/java/org/apache/activemq/artemis/core/config/WildcardConfigurationTest.java:
```diff
@@ -0,0 +1,40 @@
+public class WildcardConfigurationTest extends Assert {
+
+   public static final WildcardConfiguration MQTT_WILDCARD = new WildcardConfiguration().setDelimiter('/').setAnyWords('#').setSingleWord('+');
+   public static final WildcardConfiguration DEFAULT_WILDCARD = new WildcardConfiguration();
```
Done.

gemmellr commented on artemis-server/src/test/java/org/apache/activemq/artemis/core/config/WildcardConfigurationTest.java:
```diff
@@ -0,0 +1,60 @@
+      WildcardConfiguration a = new WildcardConfiguration().setDelimiter('/').setAnyWords('#').setSingleWord('+');
+      WildcardConfiguration b = new WildcardConfiguration().setDelimiter('/').setAnyWords('#').setSingleWord('+');
+
+      assertEquals(a, b);
```
Should also check the reverse

jbertram replied:
Done.

gemmellr commented on artemis-server/src/test/java/org/apache/activemq/artemis/core/config/WildcardConfigurationTest.java:
```diff
@@ -0,0 +1,60 @@
+      assertEquals(a, b);
+
+      String toConvert = RandomUtil.randomString();
+      assertTrue(toConvert == a.convert(toConvert, b));
+      assertTrue(toConvert == a.convert(toConvert, a));
```
Using assertSame would be nicer (tooling also tends to pick out this form of == use as a likely mistake)

jbertram replied:
Done.

jbertram commented:
@gemmellr, I just pushed a big update that hopefully addresses all the issues. Once I got into it a bit more I decided to do a more thorough refactoring of this code. There were a number of messy & inconsistent bits that I straightened out.

gemmellr commented:
@jbertram Great. I was going to skim it now, but its a bigger change now and harder to see what changed, so ill review it all again on Monday. For now, the CI job looks to be calling for some javadoc fixups.

gemmellr commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/mqtt5/MQTT5Test.java:
```diff
@@ -83,6 +83,29 @@ public void messageArrived(String topic, MqttMessage message) {
+         @Override
+         public void messageArrived(String t, MqttMessage message) {
+            logger.info("Message received from topic {}, message={}", topic, message);
+            assertEquals(topic, t);
```
Would be better to remember the value here and then assert outside the callback after the latch trips. That way the test will fail quicker and in a more helpful manner if the value isnt as expected, having the assertion directly cause it rather than e.g just timing out after 500ms on the latch (because it wasnt tripped) and requiring someone go look at logs and hope that there is a stack logged that might explain it. It would also just be clearer what is really being tested, took me a bit to even notice what the test really cared about while doing the send+receive.

jbertram replied:
Good call.

gemmellr reviewed (commented):
The latest updates seem to have caused a couple of test failures (probably around the new argument validations):
```
java.lang.IllegalArgumentException
	at org.apache.activemq.artemis.tests.integration.mqtt5.MQTT5Test.testSharedSubscriptionsWithSameName(MQTT5Test.java:359)
```

```
java.lang.IllegalArgumentException
	at org.apache.activemq.artemis.tests.integration.mqtt5.MQTT5Test.testSharedSubscriptionsWithSameName2(MQTT5Test.java:414)```

jbertram commented on artemis-protocols/artemis-mqtt-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/mqtt/MQTTUtil.java:
```diff
@@ -138,25 +138,71 @@ public class MQTTUtil {
+    */
+   public static String getCoreQueueFromMqttTopic(String topicFilter, String clientId, WildcardConfiguration wildcardConfiguration) {
+      if (topicFilter == null || clientId == null || wildcardConfiguration == null) {
+         throw new IllegalArgumentException();
```
Funny. I had already started making that change!

jbertram commented:
Those test failures should be resolved now.
